worker_tasks/categorize.py

```diff
@@ -46,109 +80,349 @@
+        links) between each pair.
         """
-        cat1, cat2 = "", ""
-        for category, weights in categories.most_common():
-            category = category.title()
-            if cat1 in cat2:
-                cat1 = category
-            elif cat2 in cat1:
-                cat2 = category
-            else:
-                break
 
-        return cat1, cat2
+        logger.debug("Building semantic graph...")
 
+        edges = []
+        vertex_attrs = {"name": []}
+        edge_attrs = {"weight": []}
 
-    def is_from_titles(self, word):
-        # Case-Insensitive Comparison
-        # since titles are uppercase and body text lowercase
-        return word.title() in self.title_categories
+        for i, article_1 in enumerate(articles):
+            vertex_attrs["name"].append(article_1)
 
+            for j, article_2 in enumerate(articles):
+                if i == j:
+                    continue
 
-    def is_bigram(self, word):
-        return " " in word
+                if (article_1, article_2) in self._memoized_semantic_relatedness_scores:
+                    relatedness_score = self._memoized_semantic_relatedness_scores[(article_1, article_2)]
+                else:
+                    relatedness_score = self._get_relatedness_score(self._memoized_article_links[article_1], self._memoized_article_links[article_2])
+                    self._memoized_semantic_relatedness_scores[(article_1, article_2)] = relatedness_score
+                    self._memoized_semantic_relatedness_scores[(article_2, article_1)] = relatedness_score
 
+                if relatedness_score:
+                    edges.append([i, j])
+                    edge_attrs["weight"].append(relatedness_score)
 
-    def is_valid_category(self, category):
+        return igraph.Graph(edges=edges, vertex_attrs=vertex_attrs, edge_attrs=edge_attrs)
 
-        # If bigram, check that both words are valid
-        if self.is_bigram(category):
-            category = category.split()
-            return self.is_valid_category(category[0]) and self.is_valid_category(category[1])
 
-        # Ignore single characters
-        if len(category) < 2:
-            return False
+    def _get_relatedness_score(self, article_1_links, article_2_links):
+        """Calculate overlap of two Wikipedia articles: 2 * [# of shared links] / [total # of links]"""
 
-        # Ignore stop words
-        if category.lower() in STOP_WORDS:
+        total_links_count = len(article_1_links) + len(article_2_links)
+
+        if not total_links_count:
+            return 0
+
+        shared_links_count = 0
+        for article_link in article_1_links:
+            weight = 1
+            if "Template" in article_link or \
+                "Wikipedia" in article_link or \
+                "Help" in article_link:
+                weight = 0.1
+            if "Category" in article_link:
+                weight = 1.5
```

Comment:
That's a good point about the Template/Wikipedia links. We may want to just make the weight 0 on those.